### PR TITLE
[Networking] Topology Tracing and Monitoring for GossipSub

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This is a PR that adds a new module called `GossipSubMeshTracer`. This module tracks the local mesh peers for each topic and logs the mesh peers and updates the local mesh size metric. It is a tracer that uses the `go-libp2p-pubsub` package to track the peers. 

The PR also includes methods for getting the mesh peers for a given topic, grafting a peer to a topic mesh, pruning a peer from a topic mesh, and logging the mesh peers at regular intervals. 

Adding the `GossipSubMeshTracer` module to the flow-go library provides insights into the local mesh peers of each node at each GossipSub topic at regular intervals. This information is essential for network debugging and optimization as it helps to identify the topology of the network.

### CODE REVIEW

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go [hunk 1]
```diff
@@ -44,8 +45,26 @@ import (
 const (
 	defaultMemoryLimitRatio     = 0.2 // flow default
 	defaultFileDescriptorsRatio = 0.5 // libp2p default
+
+	// DefaultPeerScoringEnabled is the default value for enabling peer scoring.
```
Comment (kc1116) [inline network/p2p/p2pbuilder/libp2pNodeBuilder.go:49]:
```suggestion
	// defaultPeerScoringEnabled is the default value for enabling peer scoring.
```

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go [hunk 2]
```diff
@@ -44,8 +45,26 @@ import (
...
+	defaultPeerScoringEnabled = true // enable peer scoring by default on node builder
+
+	// DefaultMeshTracerLoggingInterval is the default interval at which the mesh tracer logs the mesh
```
Comment (kc1116) [inline network/p2p/p2pbuilder/libp2pNodeBuilder.go:52]:
```suggestion
	// defaultMeshTracerLoggingInterval is the default interval at which the mesh tracer logs the mesh
```

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go [hunk 3]
```diff
@@ -44,8 +45,26 @@ import (
 const (
 	defaultMemoryLimitRatio     = 0.2 // flow default
 	defaultFileDescriptorsRatio = 0.5 // libp2p default
+
+	// DefaultPeerScoringEnabled is the default value for enabling peer scoring.
+	defaultPeerScoringEnabled = true // enable peer scoring by default on node builder
```
Comment (kc1116) [inline network/p2p/p2pbuilder/libp2pNodeBuilder.go:50]:
```suggestion
	// defaultPeerScoringEnabled is the default value for enabling peer scoring.
	// peer scoring is enabled by default.
	defaultPeerScoringEnabled = true
```

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		for p := range t.topicMeshMap[topic] {
+			id, exists := t.idProvider.ByPeerID(p)
+			if !exists {
+				shouldWarn = true
+				lg = lg.With().
+					Str("peer_id", p.String()).
+					Str("flow_id", "unknown").
+					Str("role", "unknown").
+					Logger()
+
```
Comment (kc1116) [inline network/p2p/tracer/gossipSubMeshTracer.go:165]:
```suggestion
```

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+				Str("peer_id", p.String()).
+				Hex("flow_id", logging.ID(id.NodeID)).
+				Str("role", id.Role.String()).
+				Logger()
+		}
+
+		if shouldWarn {
+			lg.Warn().
+				Bool(logging.KeySuspicious, true).
+				Msg(MeshLogIntervalMsg)
```
Comment (kc1116) [inline network/p2p/tracer/gossipSubMeshTracer.go:179]: We should log a warning different from the normal MeshLogIntervalMsg.

File: network/p2p/tracer/gossipSubMeshTracer_test.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
+package tracer_test
```
Comment (kc1116) [inline network/p2p/tracer/gossipSubMeshTracer_test.go:1]: Should we add another test case that ensures we log a warning whenever a node is not in the identity state?

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+// logPeers logs the mesh peers of the local node for each topic.
+// Note that based on GossipSub parameters, we expect to have between 6 and 12 peers in the mesh for each topic.
+// Hence, choosing a heartbeat interval in the order of minutes should be sufficient to log the mesh peers of the local node.
+// Also, note that the mesh peers are also logged reactively when a peer is added or removed from the mesh.
+func (t *GossipSubMeshTracer) logPeers() {
+	t.topicMeshMu.RLock()
+	defer t.topicMeshMu.RUnlock()
+
```
Comment (kc1116) [inline network/p2p/tracer/gossipSubMeshTracer.go:152]:
```suggestion
```

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+	id, exists := t.idProvider.ByPeerID(p)
+	if !exists {
+		lg.Warn().
+			Bool(logging.KeySuspicious, true).
+			Msg("grafted peer not found in identity provider")
+		return
+	}
+
+	lg.Info().Hex("flow_id", logging.ID(id.NodeID)).Str("role", id.Role.String()).Msg("grafted peer")
```
Comment (peterargue) [inline network/p2p/tracer/gossipSubMeshTracer.go:96]: how often is this called? will this be too noisy at `Info` level?

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+		t.topicMeshMap[topic] = make(map[peer.ID]struct{})
+	}
+	t.topicMeshMap[topic][p] = struct{}{}
+
+	id, exists := t.idProvider.ByPeerID(p)
+	if !exists {
+		lg.Warn().
+			Bool(logging.KeySuspicious, true).
+			Msg("grafted peer not found in identity provider")
+		return
```
Comment (peterargue) [inline network/p2p/tracer/gossipSubMeshTracer.go:93]: do we still need to update the metric since its size may have changed?

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+					Str("peer_id", p.String()).
+					Str("flow_id", "unknown").
+					Str("role", "unknown").
+					Logger()
+
+				continue
+			}
+
+			lg = lg.With().
+				Str("peer_id", p.String()).
```
Comment (peterargue) [inline network/p2p/tracer/gossipSubMeshTracer.go:170]: This appears to be adding `peer_id`, `flow_id`, and `role` for each of the 6-12 topic peers. a couple of questions:
1. What does zerolog do when it gets duplicate keys? append/overwrite/something else? would it make sense to use a [dictionary](https://github.com/rs/zerolog#sub-dictionary)?
2. how many characters will these log lines be? most systems have a max after which they are truncated.

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+				Logger()
+		}
+
+		if shouldWarn {
+			lg.Warn().
+				Bool(logging.KeySuspicious, true).
+				Msg(MeshLogIntervalMsg)
+			continue
+		}
+		lg.Info().Msg(MeshLogIntervalMsg)
```
Comment (peterargue) [inline network/p2p/tracer/gossipSubMeshTracer.go:182]: A couple of questions here:
1. Does this need to be logged at `Info` level, or could it be `Debug`/`Trace`? Depending on how often `logPeers` is called, this could add quite a bit of data to the logs.
2. Does it make sense to just log counts here, and move the data dump to an admin command?

File: network/p2p/tracer/gossipSubMeshTracer.go (hunk 5 above)
Comment (yhassanzadeh13) [reply network/p2p/tracer/gossipSubMeshTracer.go]: https://github.com/onflow/flow-go/pull/3970/commits/7472f44aa31e8ddfc8d37d8fa2bc89fffc5c5b7a

File: network/p2p/tracer/gossipSubMeshTracer_test.go (hunk 6 above)
Comment (yhassanzadeh13) [reply network/p2p/tracer/gossipSubMeshTracer_test.go:1]: https://github.com/onflow/flow-go/pull/3970/commits/0097d6752cd80e054ff3c917b89d77efc9e9e1c1

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go [hunk 12]
```diff
@@ -392,10 +434,20 @@ func (builder *LibP2PNodeBuilder) Build() (p2p.LibP2PNode, error) {
 					ctx.Throw(fmt.Errorf("could not stop libp2p node: %w", err))
 				}
 			}
-		}).
-		Build()
+		})
+	cm = cm.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/libp2pNodeBuilder.go:407]: https://github.com/onflow/flow-go/pull/3970/commits/da1a9277892a15e52d23bdda09c9905b0fd4ee8f

File: network/p2p/tracer/gossipSubMeshTracer.go (hunk 8 above)
Comment (yhassanzadeh13) [reply network/p2p/tracer/gossipSubMeshTracer.go:103]: The logging does not deem frequent. It is done once the startup of that node builds up its mesh and occasional changes in the local meshes of the nodes due to the peer scoring that it PRUNEs from some nodes and GRAFTs to others. Noting that the local mesh of a node cannot go beyond 12 in size so at most we have 12 * (number of channels) logs at the startup. 

This can be noisy only when a malicious peer starts spamming with GRAFTs and PRUNEs, which is detected and mitigated by our spamming prevention mechanism. Hence, overall, the impact is minimal. However, since our periodic logging intervals are designed to be long (to avoid being noisy), I think having such event-based logging upon GRAFT and PRUNEs can provide more scrutiny into what is happening between two intervals, e.g., some peers may GRAFT and PRUNE in shorter intervals than our logging interval so they less frequently are being logged. 

In conclusion, I am fine with switching to a lower-level log, but as far as I can see, it is less likely to be noisy unless a bug.

File: network/p2p/tracer/gossipSubMeshTracer.go (hunk 9 above)
Comment (yhassanzadeh13) [reply network/p2p/tracer/gossipSubMeshTracer.go:100]: Size changes only when a `GRAFT` or `PRUNE` message is sent or received. In both cases, their [relative call-back tracer methods]https://github.com/onflow/flow-go/blob/b6e255a84748b55a4de6c68053194ff9fb1332bc/network/p2p/tracer/gossipSubMeshTracer.go#L80-L132() are called by the GossipSub and the size metric is updated. The periodic logging is primarily for the situations where the topic-meshes remain the same with no new GRAFT and PRUNE sent or received. Hence, we don't have any event-based logging through `GRAFT` and `PRUNE` method calls.

File: network/p2p/tracer/gossipSubMeshTracer.go (hunk 10 above)
Comment (yhassanzadeh13) [reply network/p2p/tracer/gossipSubMeshTracer.go:178]: This code logs the local peers on each topic mesh separately, and prints a log entry per topic. The `zerolog` dictionary seems applicable to cases where we dump the entire topic mesh topology in a single log, which helps group the log entries based on the keys and merge the duplicates. I think having a separate log entry per topic (as the current approach) provides more clarity and simplifies debugging that dumping all topology-related info in the same log. 

Also, looking at a single log, the total number of characters for the keys and values in the key-value pair breakdown are:
`peer_id`: key has 7 characters, value has 52 characters
`flow_id`: key has 7 characters, value has 64 characters
`role`: key has 4 characters, value has 9 characters
Therefore, in the topic-based approach, adding each peer entry takes the total number of 143 characters, i.e., the total number of characters for the keys is 18 (7 + 7 + 4) and the total number of characters for the values is 125 (52 + 64 + 9).
Each log entry in this way also has around 70 characters for general meta-data and messages including the log level. So, with `x` peers in a topic mesh, the topic-based logging takes `70 + 125 * x`. This is around `720` characters for `6`-nodes topic meshes and around `1570` characters for `12-node` topic meshes. Accordingly, I think the current approach is safe with respect to truncation.

File: network/p2p/tracer/gossipSubMeshTracer.go (hunk 11 above)
Comment (yhassanzadeh13) [reply network/p2p/tracer/gossipSubMeshTracer.go:190]: > Does this need to be logged at Info level, or could it be Debug/Trace? Depending on how often logPeers is called, this could add quite a bit of data to the logs.

To facilitate debugging and incident response, it's helpful to have topology logs available at the Info level so that we can also have insight into the recent topology updates if any. Currently, these logs are being generated at 1-minute intervals. However, if it's found to be too noisy, we can adjust the interval to be longer, such as every 5 minutes. The logging interval can be configured by setting a flag.

> Does it make sense to just log counts here, and move the data dump to an admin command?

The admin command for topology monitoring may not be the most efficient approach for topology-related incidents. In order to have proper observability of the recent change history of the topology, as well as any real-time changes that happen to it, we need a more comprehensive solution. For instance, we may need to see whether the mesh of a particular topic had a uniform distribution of participating roles in recent history. The admin command, unfortunately, only provides a snapshot (as well as the forward history).

To overcome this limitation, the current approach provides insights into the entire topology updates of the nodes through our single logging interface. This will avoid the need to SSH into each node and run admin commands individually, which can be time-consuming and inefficient during the incidents. With a logging dashboard, we can quickly identify any anomalies or inconsistencies in the topology and take corrective actions accordingly.

**Review by kc1116 — APPROVED** [review]

File: network/p2p/tracer/gossipSubMeshTracer.go (hunk 10 above)
Comment (peterargue) [reply network/p2p/tracer/gossipSubMeshTracer.go:184]: > There will be no duplicate key on a log

I mean that there will be up to 12 `peer_id` entries in the log line right? I _think_ zerolog will allow this, and that it will serialize all labels in the order they were set, but i'm not sure. Have you tested it? Using a dictionary would make sure the groupings are preserved even if the log consumer doesn't maintain the original order.

**Review by peterargue — APPROVED** [review]

**Comment by yhassanzadeh13** [comment]
@peterargue regarding this https://github.com/onflow/flow-go/pull/3970#discussion_r1129890947:

Here is a sample of how the current logging looks like: 
```
{"level":"warn","component":"gossip_sub_topology_tracer","heartbeat_interval":1000,"topic":"push-blocks/a6ae7562fbf843ff9fd11046cf663dae5caaa9e16eabcef2129f6447aeeea805","peer_id":"16Uiu2HAmJi214g9N9irC8pnmbtwQTV6zQQiHE1iiFarLZt8iP5tq","flow_id":"7a614338c8e92678eb9a7b1593ad175f1b8ba45825f9e3a2d175b348346f490b","role":"consensus","peer_id":"16Uiu2HAkuc6ChydHsvZfeeq5b8YErSA7WyNe3CWarX1SuVPZ9UxZ","flow_id":"unknown","role":"unknown","peer_id":"16Uiu2HAmAgQUTZfkg64VKwyhYfbPBez5m7Yo9PtarWkB1brEkje4","flow_id":"b2484108e9fbb96c26406151dd651e286fc4dcb83ad56c06b5b1a9fb87b7003e","role":"consensus","suspicious":true,"message":"unknown peers in topic mesh peers of local node since last heartbeat"}

```

As shown above, for each node on a mesh, it logs the `peer_id`, `flow_id`, and `role` of the node. Hence, with `x` nodes on a mesh of a topic, we gonna have `3 * x` repeated keys. The maximum value for `x` is `12` in the current setup. We can remove the repetition by your suggested approach which groups all `peer_id`s, `flow_id`s, and `role`s together. The resulted log would look like this: 

```
{"level":"warn","component":"gossip_sub_topology_tracer","topic":"push-blocks/a6ae7562fbf843ff9fd11046cf663dae5caaa9e16eabcef2129f6447aeeea805","suspicious":true,"message":"unknown peers in topic mesh peers of local node since last heartbeat","peer_id":["16Uiu2HAmJi214g9N9irC8pnmbtwQTV6zQQiHE1iiFarLZt8iP5tq","16Uiu2HAkuc6ChydHsvZfeeq5b8YErSA7WyNe3CWarX1SuVPZ9UxZ","16Uiu2HAmAgQUTZfkg64VKwyhYfbPBez5m7Yo9PtarWkB1brEkje4"],"flow_id":["7a614338c8e92678eb9a7b1593ad175f1b8ba45825f9e3a2d175b348346f490b","unknown","b2484108e9fbb96c26406151dd651e286fc4dcb83ad56c06b5b1a9fb87b7003e"],"role":["consensus","unknown","consensus"],"heartbeat_interval":1000}
```

By grouping the repeated keys, we can save up to 18 characters per logged node. For example, with 6-12 nodes on a mesh, this can save 108-216 characters per logged message. However, this comes at the cost of reduced readability, especially when dealing with crowded topics. If this tradeoff is acceptable, I can address it in a separate PR. Separating the changes into a different PR allows for easy reversion in case the readability issue becomes problematic in the long term.

**Comment by yhassanzadeh13** [comment]
bors merge
